Implement `Encodable`/`Decodable` for list of `u8`

tcharding opened the pull request:
Currently we have a macro that implements for a vector of a bunch of types, also we implement manually for `Vec<u8>`. This was done because of the lack of specialization in Rust. However we can check the type ID of the generic type at run time and switch on it.

Implement both traits using `TypeId` to dynamically run separate code blocks for `u8` vs any other type. 

NOTE: Introduces a single line usage of `unsafe`.

apoelstra commented:
We should use Kix's trick to implement specialization on `u8` by just looking at the type id in our generic impl for `Vec<T>`.

tcharding commented:
That was too easy, as I suspected it would be, yesterday based on that assumption I thought there must be some other reason we did not do it originally.

tcharding commented:
Damn, fail. Will fix. Good eye.

apoelstra commented:
> Got a hot tip from Claude AI to use `Vec::from_raw_parts` instead of `transmute`.

Neat! But not quite clever enough.

* You have to cast a `*const` pointer to a `*mut` pointer in a way that isn't really kosher
* If `consensus_encode_with_size` or `decode_from_finite_reader` panics, the `Vec` you're so careful to forget will get dropped.

If you instead use [slice::from_raw_parts](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html) and [ptr::cast](https://doc.rust-lang.org/std/primitive.pointer.html#method.cast) which prevents const/mut changes you will avoid these problems.

apoelstra commented:
BTW `ptr::cast` is available since Rust 1.38.

`cast_const` and `cast_mut`, which I suspect are the more common case, aren't til 1.65 so we can't use them. I had thought, until right now when I looked it up, that `cast` was the same. But we should go through all of our pointer casts and check whether we can't replace them with calls to `cast`.

tcharding commented:
Done and TIL, thanks man.

tcharding commented:
Thanks @dpc, since your link included an impl for slices of `u8` I added that also.

@apoelstra recently you commented that we removed this in the past, are you happy with this logic - both `Vec<u8>` and `&[u8]` use the same logic, both include the leading compact size.

tcharding commented on bitcoin/src/consensus/encode.rs:
```diff
@@ -494,58 +490,72 @@ impl Encodable for [u16; 8] {
+            // Safe because `T` is a `u8`.
+            unsafe { 
+                // We cannot use `ptr.cast_mut` until Rust 1.65
+                Ok(Vec::from_raw_parts(ptr as *mut T, len, capacity))
```
Thanks man, I created #4797 from this reading this link.

apoelstra commented:
Is the missing impl for `&[u8]` causing us problems? Yes, if you want to justify it the change, it'd be worthwhile to go through the codebase and use it.

We don't have it right now because in order versions of Rust it was fairly unpredictable whether a `&[u8; N]` would coerce into a `&[u8]`, and because these are encoded very differently, we removed the footgun. I'd be okay putting it back if it's causing us problems, especially since Rust 2018 (or was it 2021?) that fixed a ton of array-related stuff.

Also, if we do want to do this, you should implement it on `[T]`, not on `&[T]`. There is explicitly no `Sized` bound on `Encodable`.

tcharding commented:
I've dropped the slice stuff since it is kind of out of scope and I pulled it in because I thought, incorrectly, that it would be trivial to add. Added a comment to the already existing issue to reference this PR.

tcharding commented:
Can you prioritize this one please @apoelstra, the consenus encoding stuff is on top of it.

apoelstra reviewed (approved):
ACK c2b6cf980fc8ad823993c262ed6e45c362b8c7d3; successfully ran local tests

apoelstra commented:
This business with finite readers and such is very weird. Dates to #1023 and #1360 (which link to extensive discussion as well as links to a production failure caused by our old limited parsing logic). The specific breakage is tested by #1359.

I have a more extensive unit test here:

```diff
diff --git a/bitcoin/src/consensus/encode.rs b/bitcoin/src/consensus/encode.rs
index 55f4425136..ad8cdb2cb7 100644
--- a/bitcoin/src/consensus/encode.rs
+++ b/bitcoin/src/consensus/encode.rs
@@ -1137,4 +1137,45 @@
             FromHexError::Decode(DecodeError::Unconsumed)
         ));
     }
+
+    #[test]
+    fn deserialize_extreme_tx() {
+        use crate::{ScriptBuf, Witness};
+
+        // Start with transaction from `deserialize_tx_hex`
+        let hex = include_str!("../../tests/data/previous_tx_0_hex"); // An arbitrary transaction.
+        let tx = deserialize_hex::<Transaction>(hex).unwrap();
+
+        assert_eq!(tx.input.len(), 1);
+        assert_eq!(tx.output.len(), 2);
+        assert_eq!(tx.input[0].witness.len(), 2);
+
+        // 1. Test with 4 million witnesses.
+        let mut tx_copy = tx.clone();
+        tx_copy.input[0].witness = Witness::from_slice(&vec![vec![]; 4_000_000]);
+        let roundtrip = deserialize(&serialize(&tx_copy)).unwrap();
+        assert_eq!(tx_copy, roundtrip);
+
+        // 2. Test with a single large witness. (Size of 4 megs, including length prefix)
+        let mut tx_copy = tx.clone();
+        tx_copy.input[0].witness = Witness::from_slice(&vec![vec![0; 4_000_000 - 9]; 1]);
+        let roundtrip = deserialize(&serialize(&tx_copy)).unwrap();
+        assert_eq!(tx_copy, roundtrip);
+
+        // 3. Combine these; with the witness stack we can exceed a total size of 4M but
+        //    only by a tiny bit. (It is not part of our API guarantee that such things
+        //    will round-trip, but we unit test them anyway to help notice changes.)
+        let mut tx_copy = tx.clone();
+        tx_copy.input[0].witness = Witness::from_slice(&vec![vec![0; 997]; 4_000]);
+        let roundtrip = deserialize(&serialize(&tx_copy)).unwrap();
+        assert_eq!(tx_copy, roundtrip);
+
+        // 4. Test with a large script sig. With scriptsigs there is no limit on how large
+        //    an object we can parse, which is inconsistent with witnesses. Also not an
+        //    API guarantee.
+        let mut tx_copy = tx.clone();
+        tx_copy.input[0].script_sig = ScriptBuf::from(vec![0; 8_000_001]);
+        let roundtrip = deserialize(&serialize(&tx_copy)).unwrap();
+        assert_eq!(tx_copy, roundtrip);
+    }
 }
```

which I will PR separately. Confirmed that it passes before and after this PR, though as you can see it reveals behavior which is kinda weird around oversized transactions.
